docker/generate_makefile.py

```diff
@@ -4,200 +4,140 @@
-    -e FUZZ_TARGET=$({benchmark}-fuzz-target) \\
-    --entrypoint "/bin/bash" \\
-    -it {base_tag}/runners/{fuzzer}/{benchmark}
-"""
-
-FUZZER_BENCHMARK_TEMPLATE = """
-.{fuzzer}-{benchmark}-builder-intermediate: .{benchmark}-project-builder
-	docker build \\
-    --tag {base_tag}/builders/{fuzzer}/{benchmark}-intermediate \\
-    --file=fuzzers/{fuzzer}/builder.Dockerfile \\
-    --build-arg parent_image=gcr.io/fuzzbench/builders/benchmark/{benchmark} \\
-    $(call cache_from,{base_tag}/builders/{fuzzer}/{benchmark}-intermediate) \\
-    fuzzers/{fuzzer}
-
-.pull-{fuzzer}-{benchmark}-builder-intermediate:
-	docker pull {base_tag}/builders/{fuzzer}/{benchmark}-intermediate
-
-.{fuzzer}-{benchmark}-builder: .{fuzzer}-{benchmark}-builder-intermediate
-	docker build \\
-    --tag {base_tag}/builders/{fuzzer}/{benchmark} \\
-    --file=docker/benchmark-builder/Dockerfile \\
-    --build-arg parent_image={base_tag}/builders/{fuzzer}/{benchmark}-intermediate \\
-    --build-arg fuzzer={fuzzer} \\
-    --build-arg benchmark={benchmark} \\
-    $(call cache_from,{base_tag}/builders/{fuzzer}/{benchmark}) \\
-    .
-
-.pull-{fuzzer}-{benchmark}-builder: .pull-{fuzzer}-{benchmark}-builder-intermediate
-	docker pull {base_tag}/builders/{fuzzer}/{benchmark}
-
-ifneq ({fuzzer}, coverage)
-
-.{fuzzer}-{benchmark}-intermediate-runner: base-runner
-	docker build \\
-    --tag {base_tag}/runners/{fuzzer}/{benchmark}-intermediate \\
-    --file fuzzers/{fuzzer}/runner.Dockerfile \\
-    $(call cache_from,{base_tag}/runners/{fuzzer}/{benchmark}-intermediate) \\
-    fuzzers/{fuzzer}
-
-.pull-{fuzzer}-{benchmark}-intermediate-runner: pull-base-runner
-	docker pull {base_tag}/runners/{fuzzer}/{benchmark}-intermediate
-
-.{fuzzer}-{benchmark}-runner: .{fuzzer}-{benchmark}-builder .{fuzzer}-{benchmark}-intermediate-runner
-	docker build \\
-    --tag {base_tag}/runners/{fuzzer}/{benchmark} \\
-    --build-arg fuzzer={fuzzer} \\
-    --build-arg benchmark={benchmark} \\
-    $(call cache_from,{base_tag}/runners/{fuzzer}/{benchmark}) \\
-    --file docker/benchmark-runner/Dockerfile \\
-    .
-
-.pull-{fuzzer}-{benchmark}-runner: .pull-{fuzzer}-{benchmark}-builder .pull-{fuzzer}-{benchmark}-intermediate-runner
-	docker pull {base_tag}/runners/{fuzzer}/{benchmark}
-
-""" + FUZZER_BENCHMARK_RUN_TARGETS_TEMPLATE + """
-
-else
-
-build-{fuzzer}-{benchmark}: .{fuzzer}-{benchmark}-builder
-pull-{fuzzer}-{benchmark}: .pull-{fuzzer}-{benchmark}-builder
-
-endif
-"""
-
-
-def generate_fuzzer(fuzzer, benchmarks):
-    """Output make rules for a single fuzzer."""
-    # Generate build rules for the fuzzer itself.
-    print(FUZZER_TEMPLATE.format(fuzzer=fuzzer, base_tag=BASE_TAG))
-
-    # Generate rules for fuzzer-benchmark pairs.
+from common import yaml_utils
+from common import benchmark_utils
+from common import fuzzer_utils
+from experiment.build import docker_images
+
+BASE_TAG = "gcr.io/fuzzbench"
+BENCHMARK_DIR = benchmark_utils.BENCHMARKS_DIR
+
+RUN_TEMPLATE = """
```

Comment:
it's difficult to see what we're generating when we have a template for this part here, and we generate the other parts in `_print_makefile_build_template()` with print statements. it'd be easier to read if everything would be at one place. to do that consider either 

1) do everything in `_print_makefile_build_template()` in the same fashion, or 
2) do everything with a jinja template  (you'll need jinja to implement if-s and for-s)

Reply:
I used the RUN_TEMPLATE because its very small. If I print the run commands line-wise like in `_print_makefile_build_template()`, but in `_print_makefile_run_template()`, would that be acceptable? wdyt? Or as you said for (1), print the run commands also with the same function?

IMO, these python conditionals are easier to read than a single jinja template specially considering nested for and if statements. I think jinja template won't look less convoluted from current master's FUZZER_TEMPLATE.

Reply:
> I used the RUN_TEMPLATE because its very small. If I print the run commands line-wise like in `_print_makefile_build_template()`, but in `_print_makefile_run_template()`, would that be acceptable? wdyt? 

yes, sure, thanks!

(i'd personally probably just have a single `print_rules_for_image(name, image)` function and do everything there, but two functions next each other is fine with me too :) )

Reply:
Done